### PR TITLE
Feature/addSchema to serialization

### LINKED ISSUE #1410: Serialization references not resolved
## 🐛 Bug Report

References for validation can be added using the `addSchema` method, so in our JSON schemas we can safely use `$ref` to reference them.

However, when trying to do the same for the serialization using `fast-json-stringify` I get the following error:

```
Error: schema is invalid: data.items.$ref should be string, data.items should be array, data.items should match some schema in anyOf
```

Reading up on the documentation (https://github.com/fastify/fast-json-stringify#ref) it seems that external definitions need to be passed as an option, so I was wondering if the `addSchema` function would not already do this?

## To Reproduce

asset.schema.json
```json
{
  "$id": "http://poppy.be/schemas/asset.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Physical Asset",
  "description": "A generic representation of a physical asset",
  "type": "object",
  "required": [
    "id",
    "model",
    "location"
  ],
  "properties": {
    "id": {
      "type": "string",
      "format": "uuid"
    },
    "model": {
      "type": "string"
    },
    "location": { "$ref": "http://poppy.be/schemas/geo/point.json#" }
  }
}
```

point.schema.json
```json
{
  "$id": "http://poppy.be/schemas/geo/point.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Longitude and Latitude Values",
  "description": "A geographical coordinate.",
  "type": "object",
  "required": [
    "latitude",
    "longitude"
  ],
  "properties": {
    "latitude": {
      "type": "number",
      "minimum": -90,
      "maximum": 90
    },
    "longitude": {
      "type": "number",
      "minimum": -180,
      "maximum": 180
    },
    "altitude": {
      "type": "number"
    }
  }
}
```

assetResponse.schema.json
```json
{
  "$id": "http://poppy.be/schemas/response/assetLocations.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "List of Asset locations",
  "type": "array",
  "items": { "$ref": "http://poppy.be/schemas/asset.json#" },
  "default": []
}
```

Route
```js
app.get('/assets', { schema: { querystring: Schemas.geo.point, response: { 200: Schemas.response.assetLocations } } }, async (req, res) => {
  ...
})
```

## Expected behavior

The serialization should simply be able to resolve the `asset` schema.

## Your Environment

- *node version*: 10
- *fastify version*: >=1.0.0
- *os*: Mac

### PR DESCRIPTION
**Pull request by Eomm**
Hi,
this PR fixes #1410  and add more tests for some use cases of JSON-schema.

Regarding the table for the compatibility, I have tested the not covered use cases (I will open an issue to implement them. [Here a GIST](https://gist.github.com/Eomm/2e6e01a02e06d34256c767e076b5562f#file-test-validator-ref-id) as test)

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Validation-and-Serialization.md [hunk 1]
```diff
@@ -302,6 +303,20 @@ fastify.setErrorHandler(function (error, request, reply) {
...
+JSON Schema has some type of utilities in order to optimize your schemas that,
+in conjuction with the Fastify's shared schema, let you reuse all your schemas easily.
+
+| Use Case                          | Validator | Serializer |
+|-----------------------------------|-----------|------------|
+| shared schema                     | ✔️ | ✔️ |
+| `$ref` to `$id`                   | ✔ | ✔️ |
+| `$ref` to `/definitions`          | ✔️ | ✔️ |
+| `$ref` to shared schema `$id`          | ❌ | ✔️ |
+| `$ref` to shared schema `/definitions` | ❌ | ✔️ |
```
Comment (mcollina) [inline docs/Validation-and-Serialization.md]: I think it would be good to add some examples

Comment (Eomm) [reply docs/Validation-and-Serialization.md]: Done 👍

**Comment by Eomm** [comment]
Yes!

Because I filtered only the `$ref` key that can't use the shared schema feature but it must be a string without being `resolved` by the `schemas.js` class.

**Comment by mcollina** [comment]
Can you send a backport PR to 1.x with https://github.com/fastify/fastify/commit/972f2e7b124d4ca692c25c1c4b1be46c6913c0e9?

**Comment by Eomm** [comment]
Sure 👍

**Comment by delvedor** [comment]
Is this a semver minor or a bugfix?

**Comment by Eomm** [comment]
It is a semver patch.

The PR title is misleading 😅

**Pull request merged**